### PR TITLE
kafka replay speed: add alert for when we miss records in Kafka

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
#### What this PR does

Adds an alert and metrics to detect when we have bugs.

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: development/mimir-ingest-storage/docker-compose.jsonnet [hunk 1]
```diff
@@ -54,8 +54,6 @@ std.manifestYamlDoc({
         '-ingester.ring.prefix=exclusive-prefix',
         '-ingest-storage.kafka.consume-from-position-at-startup=end',
         '-ingest-storage.kafka.consume-from-timestamp-at-startup=0',
-        '-ingest-storage.kafka.ingestion-concurrency=2',
```
Comment (pracucci) [inline development/mimir-ingest-storage/docker-compose.jsonnet:57]: What this change intentional, or should we keep it enabled in the local dev env, just changing the config option to `...-max`?

**Review by pracucci — APPROVED** [review]
Good job. Can you add an assertion to existing unit tests to ensure the metric is always 0 at the end of each test? Should be quick to do.

File: development/mimir-ingest-storage/docker-compose.jsonnet (hunk 1 above)
Comment (dimitarvdimitrov) [reply development/mimir-ingest-storage/docker-compose.jsonnet:57]: that was a mistake. I restored them in this commit 4b64ae6f3ca5f31aa15332d508a069444115fd40

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 2]
```diff
@@ -1529,6 +1529,24 @@ How to **investigate**:
   - If the call exists and it's waiting on a lock then there may be a deadlock.
   - If the call doesn't exist then it could either mean processing is not stuck (false positive) or the `pushToStorage` wasn't called at all, and so you should investigate the callers in the code.
 
+### MimirIngesterMissedRecordsFromKafka
+
+This alert fires when an ingester has missed processing some records from Kafka. In other words, there has been a gap in offsets.
+
+How it **works**:
+
+- Ingester reads records from Kafka, and processes them sequentially. It keeps track of the offset of the last record it processed.
```
Comment (tacole02) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:1538]:
```suggestion
- The ingester reads records from Kafka and processes them sequentially. It keeps track of the offset of the last record it's processed.
```

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 3]
```diff
@@ -1529,6 +1529,24 @@ How to **investigate**:
...
+- Upon fetching the next batch of records, it checks if the first available record has an offset one greater than the last processed offset. If the first available offset is larger than that, then the ingester has missed some records.
```
Comment (tacole02) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:1539]:
```suggestion
- Upon fetching the next batch of records, it checks if the first available record has an offset of one greater than the last processed offset. If the first available offset is larger than that, then the ingester has missed some records.
```

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 4]
```diff
@@ -1529,6 +1529,24 @@ How to **investigate**:
...
+- Kafka doesn't guarantee sequential offsets. If a record has been manually deleted from Kafka or the records have been produced in a transaction and the transaction was aborted, then there may be a gap.
```
Comment (tacole02) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:1540]:
```suggestion
- Kafka doesn't guarantee sequential offsets. If a record has been manually deleted from Kafka or if the records have been produced in a transaction and the transaction was aborted, then there may be a gap.
```

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 5]
```diff
@@ -1529,6 +1529,24 @@ How to **investigate**:
...
+- Mimir doesn't produce in transactions and does not delete records.
+- When the ingester starts up, it will attempt to resume from the last offset it processed. If the ingester has been unavailable for long enough that the next record is already removed due to retention, then the ingester will miss some records.
```
Comment (tacole02) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:1542]:
```suggestion
- When the ingester starts, it attempts to resume from the last offset it processed. If the ingester has been unavailable for long enough that the next record is already removed due to retention, then the ingester misses some records.
```

Comment (tacole02) [reply docs/sources/mimir/manage/mimir-runbooks/_index.md:1542]: We avoid using future tense in the docs.

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 6]
```diff
@@ -1529,6 +1529,24 @@ How to **investigate**:
...
+### MimirIngesterMissedRecordsFromKafka
+
+This alert fires when an ingester has missed processing some records from Kafka. In other words, there has been a gap in offsets.
+
+How it **works**:
+
+- Ingester reads records from Kafka, and processes them sequentially. It keeps track of the offset of the last record it processed.
+- Upon fetching the next batch of records, it checks if the first available record has an offset one greater than the last processed offset. If the first available offset is larger than that, then the ingester has missed some records.
+- Kafka doesn't guarantee sequential offsets. If a record has been manually deleted from Kafka or the records have been produced in a transaction and the transaction was aborted, then there may be a gap.
+- Mimir doesn't produce in transactions and does not delete records.
```
Comment (tacole02) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:1541]: "Mimir doesn't produce in transactions" reads unclear to me. Is the "in" supposed to be here?

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 7]
```diff
@@ -1529,6 +1529,24 @@ How to **investigate**:
...
+- When the ingester starts up, it will attempt to resume from the last offset it processed. If the ingester has been unavailable for long enough that the next record is already removed due to retention, then the ingester will miss some records.
+
+How to **investigate**:
+
+- Verify that there have been no deleted records in your Kafka cluster by humans or other applications.
```
Comment (tacole02) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:1546]: I think we can probably remove "by humans or other applications".

**Review by tacole02 — APPROVED** [review]
Looks good! A few minor questions/suggestions.

File: docs/sources/mimir/manage/mimir-runbooks/_index.md (hunk 6 above)
Comment (dimitarvdimitrov) [reply docs/sources/mimir/manage/mimir-runbooks/_index.md:1541]: yes. in Kafka you can create a transaction and in its context produce records.

File: docs/sources/mimir/manage/mimir-runbooks/_index.md (hunk 7 above)
Comment (dimitarvdimitrov) [reply docs/sources/mimir/manage/mimir-runbooks/_index.md:1546]:
```suggestion
- Verify that there have been no deleted records in your Kafka cluster.
```

File: pkg/storage/ingest/fetcher_test.go [hunk 8]
```diff
@@ -1149,6 +1151,15 @@ func createConcurrentFetchers(ctx context.Context, t *testing.T, client *kgo.Cli
 	reg := prometheus.NewPedanticRegistry()
 	metrics := newReaderMetrics(partition, reg, noopReaderMetricsSource{})
 
+	t.Cleanup(func() {
```
Comment (pracucci) [inline pkg/storage/ingest/fetcher_test.go:1154]: Nice use of cleanup!

File: pkg/storage/ingest/fetcher.go [hunk 9]
```diff
@@ -388,6 +390,41 @@ func (r *concurrentFetchers) PollFetches(ctx context.Context) (kgo.Fetches, cont
 	}
 }
 
+func instrumentGaps(gaps []offsetRange, records prometheus.Counter, logger log.Logger) {
+	for _, gap := range gaps {
+		level.Error(logger).Log(
+			"msg", "there is a gap in consumed offsets; it is likely that there was data loss; see runbook for MimirIngesterMissedRecordsFromKafka",
+			"gap_start_inclusive", gap.start,
+			"gap_end_inclusive", gap.end,
```
Comment (pracucci) [inline pkg/storage/ingest/fetcher.go:398]: The `offsetRange` doc says that `end` is exclusive:

```suggestion
			"records_offset_gap_end_exclusive", gap.end,
```

File: pkg/storage/ingest/fetcher.go [hunk 10]
```diff
@@ -388,6 +390,41 @@ func (r *concurrentFetchers) PollFetches(ctx context.Context) (kgo.Fetches, cont
 	}
 }
 
+func instrumentGaps(gaps []offsetRange, records prometheus.Counter, logger log.Logger) {
+	for _, gap := range gaps {
+		level.Error(logger).Log(
+			"msg", "there is a gap in consumed offsets; it is likely that there was data loss; see runbook for MimirIngesterMissedRecordsFromKafka",
+			"gap_start_inclusive", gap.start,
```
Comment (pracucci) [inline pkg/storage/ingest/fetcher.go:397]: [nit] Just to clarify what's that.

```suggestion
			"records_offset_gap_start_inclusive", gap.start,
```

File: pkg/storage/ingest/fetcher.go [hunk 11]
```diff
@@ -388,6 +390,41 @@ func (r *concurrentFetchers) PollFetches(ctx context.Context) (kgo.Fetches, cont
...
+
+	// end is exclusive
+	end int64
+}
+
+func (g offsetRange) numOffsets() int64 {
+	return g.end - g.start
+}
+
+func findGapsInRecords(records []*kgo.Record, lastReturnedOffset int64) []offsetRange {
```
Comment (pracucci) [inline pkg/storage/ingest/fetcher.go:417]: Can you add a simple unit test before merging, please?

**Review by pracucci — APPROVED** [review]
Nice job!

File: docs/sources/mimir/manage/mimir-runbooks/_index.md [hunk 12]
```diff
@@ -1529,6 +1529,24 @@ How to **investigate**:
...
+
+- The ingester reads records from Kafka and processes them sequentially. It keeps track of the offset of the last record it's processed.
+- Upon fetching the next batch of records, it checks if the first available record has an offset of one greater than the last processed offset. If the first available offset is larger than that, then the ingester has missed some records.
+- Kafka doesn't guarantee sequential offsets. If a record has been manually deleted from Kafka or if the records have been produced in a transaction and the transaction was aborted, then there may be a gap.
+- Mimir doesn't produce in transactions and does not delete records.
+- When the ingester starts, it attempts to resume from the last offset it processed. If the ingester has been unavailable for long enough that the next record is already removed due to retention, then the ingester misses some records.
+
+How to **investigate**:
+
+- Verify that there have been no deleted records in your Kafka cluster.
```
Comment (dimitarvdimitrov) [inline docs/sources/mimir/manage/mimir-runbooks/_index.md:1546]:
```suggestion
- Find the offsets which were missed. The ingester logs them along with the message `there is a gap in consumed offsets`.
- Verify that there have been no deleted records in your Kafka cluster.
```

File: pkg/storage/ingest/fetcher.go (hunk 11 above)
Comment (dimitarvdimitrov) [reply pkg/storage/ingest/fetcher.go:417]: done here 5ed7788dc49d2c077a9175085e465b55f60b1454

**Pull request merged**